src/legacy/ui/public/index_patterns/_index_pattern.ts

```diff
@@ -0,0 +1,535 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import _, { each, reject } from 'lodash';
+import { i18n } from '@kbn/i18n';
+// @ts-ignore
+import { SavedObjectNotFound, DuplicateField } from 'ui/errors';
+// @ts-ignore
+import { fieldFormats } from 'ui/registry/field_formats';
+// @ts-ignore
+import { expandShorthand } from 'ui/utils/mapping_setup';
+import { toastNotifications } from 'ui/notify';
+import { findObjectByTitle, SavedObjectsClient } from 'ui/saved_objects';
+
+import { IndexPatternsApiClient } from 'ui/index_patterns/index_patterns_api_client';
+import { IndexPatternMissingIndices } from './errors';
+import { getRoutes } from './get_routes';
+import { FieldList } from './_field_list';
+import { createFieldsFetcher } from './fields_fetcher';
+import { Field, FieldType } from './_field';
+// @ts-ignore
+import { flattenHitWrapper } from './_flatten_hit';
+// @ts-ignore
+import { formatHitProvider } from './_format_hit';
+
+const MAX_ATTEMPTS_TO_RESOLVE_CONFLICTS = 3;
+const type = 'index-pattern';
+
+interface FieldMappingSpec {
+  _serialize: (mapping: any) => string;
+  _deserialize: (mapping: string) => any;
+}
+
+interface MappingObject {
+  [key: string]: FieldMappingSpec;
+}
+
+export interface StaticIndexPattern {
```

Comment:
I believe `StaticIndexPattern` is meant to be the interface representing `IndexPattern`, so that the two can be exported separately. (If you import `IndexPattern` you are actually getting the TS class, and in many cases -- especially tests -- people aren't necessarily going to need the class for this type, but rather a separate interface).

The fact that `IndexPattern` now implements `StaticIndexPattern` is only here to make this relationship clearer.

I think we should probably be renaming these though; but it depends on the naming conventions we think we should use -- I'll bring up this topic offline for us to discuss later in the week.